macaroon-thought.md

```diff
@@ -0,0 +1,399 @@
+# Integral Principles of the Structural Dynamics of Macaroons
+
+## Describe a Fly Macaroon In The Fewest Words Possible
+
+A Macaroon is an HMAC-authenticated access token. Fly Macaroons are encoded 
+in MsgPack, which is a kind of binary JSON. Macaroons are like other kinds of access tokens, with these quirks:
+
+* They're cryptographically secured using a minimal, non-negotiable set of
+  primitives: HMAC and an AEAD cipher. There are no crypto parameters to set
+  or verify.
+  
+* They're rigidly structured: every Macaroon is a set of restrictions on 
+  what its holder can do. These are called caveats. All caveats 
+  must clear for the token to authorize an action. 
+  
+* Anybody can take a Macaroon and add additional caveats to it before
+  saving it or passing it on to someone else. This is called "attenuation". 
+  Attenuating a token can't possibly give it more powers --- once added,
+  caveats can't be removed (a clever chaining HMAC construction guarantees
+  this), and all caveats must clear.
+  
+* A special kind of "third party" (3P) caveat encodes whether some other,
+  possibly unrelated service authorizes a request. That service needs 
```

Comment:
This isn't quite accurate, but brevity may outweigh precision in this intro section. Maybe something like:

> A special kind of "third party" (3P) caveat requires some other, possibly unrelated service to authorize the user before the token can be used.